### PR TITLE
[determinism] Add softmax/cross-entropy op exceptions for GPU determinism

### PR DESCRIPTION
**Pull request by duncanriach**
## High-Level Summary

This current PR adds and tests the following functionality:

When the environment variable `TF_DETERMINISTIC_OPS` is set to `"true"` or `"1"`, an attempt to run the following ops on a GPU will throw `tf.errors.UnimplementedError` (with an understandable message).

`tf.nn.softmax_cross_entropy_with_logits`
`tf.nn.sparse_softmax_cross_entropy_with_logits`

Please see _RFC: Enhancing determinism in TF_ (being added via tensorflow/community PR [346](https://github.com/tensorflow/community/pull/346)).

### Data Types

The exceptions will be thrown for all currently GPU-supported data types for the `logits` input: `tf.float16` and `tf.float32` for both ops, and, additionally, `tf.float64` for `tf.nn.softmax_cross_entropy_with_logits`.

Exception-throwing for all combinations of relevant data types for `logits` and `labels` (`tf.int32` and `tf.int64`) are tested in both eager and graph mode when the op is used in the forward direction.

### Forward vs Backward

It is currently suspected that the introduction of random noise into the gradients passed backwards from this op actually originate in the forward path algorithm, but the backward path algorithm might add additional noise. However, the backprop path for this op is not, and cannot, be used without the forward path algorithm also being used (due to this being a loss function). Therefore, the presence of exception-throwing on the backward path specifically is not necessary and is not implemented or tested by this current PR.

When these ops have a fully deterministic mode of operation, the bit-exact reproducibility of the outputs of both the forward and backward paths of the ops should be verified.

### XLA

The tests will not be run with XLA auto-jit enabled because any XLA implementation of these ops will not throw these exceptions.

When a fully deterministic mode for these ops is implemented, the bit-exact reproducibility of the outputs of both the forward and backward paths of the ops should be verified both with and without XLA auto-jit enabled.

### CODE REVIEW

**Comment by duncanriach** [comment]
@sanjoy: It would be so cool for this one to get into TensorFlow version 2.5 as well. :-)

File: tensorflow/core/kernels/xent_op.cc [hunk 1]
```diff
@@ -58,6 +82,17 @@ class SoftmaxXentWithLogitsOp : public OpKernel {
                                         "2-dimensional, or broadcasted to be "
                                         "2-dimensional"));
 
+    if (std::is_same<Device, GPUDevice>::value) {
+      OP_REQUIRES(
+          context,
+          (!RequireDeterminism() ||
+            DisableSoftmaxXentWithLogitsOpDeterminismExceptions()),
```
Comment (sanjoy) [inline tensorflow/core/kernels/xent_op.cc:89]: Unnecessary parens.

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Review by sanjoy — APPROVED** [review]

File: tensorflow/core/kernels/xent_op.cc [hunk 2]
```diff
@@ -58,6 +86,17 @@ class SoftmaxXentWithLogitsOp : public OpKernel {
                                         "2-dimensional, or broadcasted to be "
                                         "2-dimensional"));
 
+    if (std::is_same<Device, GPUDevice>::value) {
```
Comment (sanjoy) [inline tensorflow/core/kernels/xent_op.cc:89]: Is the CPU implementation deterministic?

**Review by sanjoy — APPROVED** [review]

File: tensorflow/core/kernels/xent_op.cc (hunk 2 above)
Comment (duncanriach) [reply tensorflow/core/kernels/xent_op.cc:89]: We just confirmed that the CPU implementation is deterministic (thanks @wenscarl). I'm considering add tests to prove/confirm/ensure that in a future PR.

**Comment by duncanriach** [comment]
@sanjoy, will it be possible to get this merged before the version 2.5 branch is cut, which I believe will be on March 25 (tomorrow)?

**Review by sanjoy — APPROVED** [review]

**Comment by sanjoy** [comment]
> @sanjoy, please will it be possible to get this merged before the version 2.5 branch is cut, which I believe will be on March 25 (tomorrow)?

I just approved it, but I can't guarantee that it will make it through the merge process before the branch cut.

**Comment by duncanriach** [comment]
My spam filter was acting up and I missed your comment, @sanjoy, sorry. I also missed the internal checks failure. This PR didn't make it into 2.5. Please will you let me know what the `feedback/copybara` problem is? (I can't see into it).

**Comment by sanjoy** [comment]
> My spam filter was acting up and I missed your comment, @sanjoy, sorry. I also missed the internal checks failure. This PR didn't make it into 2.5. Please will you let me know what the `feedback/copybara` problem is? (I can't see into it).

Looks like some internal tooling failure, I'll merge it manually.  Unfortunately this won't make 2.5, as you noted.
